### PR TITLE
Enable switching cameras

### PR DESCRIPTION
**Pull request by Gemakk**
## Purpose

* Adds the ability to switch front/back cameras

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

* Test the code
Verify that the following are valid
* On real and emulated devices, verify that the switch camera button is available when camera is enabled
* Verify that you can switch from the front and back cameras without issues
* **NOTE:** On Android Emulators, when you switch to the back camera, an extra tap is required afterwards, since focus was brought away from the emulator to show you that you can pan the back camera.  You won't find this same behaviour on a real device.

### CODE REVIEW

**Comment by lpx1989** [comment]
Tested on the pixel 4a, the camera switch works on all different cases of layout during the call (only self screen, 1 on 1 and grid). 
One thing I am not sure, if we need to align the upper boundary of camera switch button with the upper boundary of participants' count notification bar or not. 

Screenshot_20210518-144606

**Comment by Gemakk** [comment]
> Tested on the pixel 4a, the camera switch works on all different cases of layout during the call (only self screen, 1 on 1 and grid).
> One thing I am not sure, if we need to align the upper boundary of camera switch button with the upper boundary of participants' count notification bar or not.
> 
> Screenshot_20210518-144606

@waynemo What do you think?  I can align it when it appears, but it may throw users off when the button jumps location.  Or I can move it down a bit in a static fashion.  I suspect this button will be covered up by the compliance messages in Teams interop.

**Comment by Gemakk** [comment]
As per https://skype.visualstudio.com/SPOOL/_workitems/edit/2407134/ I've placed the button on the right center now

**Comment by Gemakk** [comment]
Thanks to @cheukchuen's feedback, I've put the button back to the top right only in the PIP window.

File: AzureCalling/app/src/main/java/com/azure/samples/communication/calling/external/calling/CallingContext.java [hunk 1]
```diff
@@ -216,6 +220,23 @@ public CompletableFuture turnOffVideoAsync() {
...
+    public CompletableFuture<LocalVideoStream> switchCameraAsync() {
+        return getLocalVideoStreamCompletableFuture().thenCompose(localVideoStream -> {
+            final VideoDeviceInfo currentCamera = localVideoStream.getSource();
+            localVideoStreamCompletableFuture = new CompletableFuture<>();
+            final LocalVideoStream newVideoStream;
+            if (currentCamera.getCameraFacing().name().equalsIgnoreCase(CameraType.FRONT.name())) {
+                newVideoStream = new LocalVideoStream(getBackCamera(), appContext);
+            } else {
+                newVideoStream = new LocalVideoStream(getFrontCamera(), appContext);
+            }
```
Comment (cheukchuen) [inline AzureCalling/app/src/main/java/com/azure/samples/communication/calling/external/calling/CallingContext.java:232]: Not sure if Android Calling SDK has the `.switchSource` method for `LocalVideoStream`. In iOS we're able to leverage `localVideoStream.switchSource(camera:completionHandler:)` to perform camera switching without the need to set a new LocalVideoStream instance

Comment (Gemakk) [reply AzureCalling/app/src/main/java/com/azure/samples/communication/calling/external/calling/CallingContext.java]: Thanks, I see Android SDK has the same thing.  I've implemented this functionality now in the latest version.

File: AzureCalling/app/src/main/java/com/azure/samples/communication/calling/activities/CallActivity.java [hunk 2]
```diff
@@ -233,7 +233,9 @@ private void initParticipantViews() {
         localParticipantView = new ParticipantView(this);
         localParticipantView.setDisplayName(callingContext.getDisplayName() + " (Me)");
         localParticipantView.setVideoDisplayed(callingContext.getCameraOn());
+        localParticipantView.setSwitchCameraButtonDisplayed(callingContext.getCameraOn());
         localParticipantView.setIsMuted(!callingContext.getMicOn());
+        localParticipantView.setImageButtonOnClickAction(this::switchCameraAsync);
```
Comment (iaulakh) [inline AzureCalling/app/src/main/java/com/azure/samples/communication/calling/activities/CallActivity.java:238]: do we need to unsubscribe on Destroy?

Comment (Gemakk) [reply AzureCalling/app/src/main/java/com/azure/samples/communication/calling/activities/CallActivity.java:238]: > do we need to unsubscribe on Destroy?

`localParticipantView` is only created once, and so setting the OnClickAction is also called once.  Additionally, this just sets the Runnable, so this particular method running multiple times is idempotent.

File: AzureCalling/app/src/main/java/com/azure/samples/communication/calling/external/calling/CallingContext.java [hunk 3]
```diff
@@ -271,28 +293,64 @@ private void createDeviceManager() {
...
+
+    private void addVideoDevices(final List<VideoDeviceInfo> addedVideoDevices) {
+        Log.d(LOG_TAG, "Added Cameras: " + addedVideoDevices.size());
+        for (final VideoDeviceInfo addedVideoDevice: addedVideoDevices) {
+            if (addedVideoDevice.getCameraFacing().name().equalsIgnoreCase(CameraType.FRONT.name())) {
+                availableCameras.put(CameraType.FRONT, addedVideoDevice);
+            } else if (addedVideoDevice.getCameraFacing().name().equalsIgnoreCase(CameraType.BACK.name())) {
+                availableCameras.put(CameraType.BACK, addedVideoDevice);
+            }
+        }
```
Comment (iaulakh) [inline AzureCalling/app/src/main/java/com/azure/samples/communication/calling/external/calling/CallingContext.java:343]: is it possible to have multiple front or back cameras?

Comment (Gemakk) [reply AzureCalling/app/src/main/java/com/azure/samples/communication/calling/external/calling/CallingContext.java:343]: Many phones nowadays have multiple front and back cameras.  I'm relying on the phone itself declaring an explicit `FRONT` or `BACK` camera for me to decide which would be the best ones to use as the 'official' front and back cameras.

**Review by iaulakh — APPROVED** [review]

**Comment by iaulakh** [comment]
tested yesterday during call with Yukon

**Review by pavelprystinka — APPROVED** [review]

**Pull request merged**